File: 2020/6/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	input := strings.Split(string(data), "\n")
	fmt.Println(part1(input))
	fmt.Println(part2(input))
}

func findAnswers(data []string) map[rune]int {
	answers := map[rune]int{}
	for _, line := range data {
		for _, char := range line {
			answers[char]++
		}
	}
	return answers
}

func part1(input []string) int {
	row := 0
	sum := 0
	for row < len(input) {
		data := []string{}
		for row < len(input) && len(input[row]) > 0 {
			data = append(data, input[row])
			row++
		}
		sum += len(findAnswers(data))
		row++
	}
	return sum
}

func part2(input []string) int {
	row := 0
	sum := 0
	for row < len(input) {
		data := []string{}
		for row < len(input) && len(input[row]) > 0 {
			data = append(data, input[row])
			row++
		}
		answers := findAnswers(data)
		for _, n := range answers {
			if n == len(data) {
				sum++
			}
		}
		row++
	}
	return sum
}

File: 2020/6/solution.py
from collections import defaultdict
import os

def part1(lines: list[str]) -> int:
  tmp = ''
  count = 0
  for line in lines:
    if not line:
      yeses = set(list(tmp))
      count += len(yeses)
      tmp = ''
    else:
      tmp += line
  yeses = set(list(tmp))
  return count + len(yeses)

def part2(lines: list[str]) -> int:
  count = 0
  yeses = defaultdict(int)
  members = 0
  for line in lines:
    if not line:
      count += sum(1 for n in yeses.values() if n == members)
      yeses = defaultdict(int)
      members = 0
    else:
      members += 1
      for c in line:
        yeses[c] += 1
  count += sum(1 for n in yeses.values() if n == members)
  return count

if __name__ == "__main__":
  with open(os.path.join(os.path.dirname(os.path.abspath(__file__)), 'input.txt'),'r') as file:
    lines = [line.strip() for line in file]

  print(part1(lines))
  print(part2(lines))
